src/resolve/resolved-mdns.c

```diff
@@ -173,12 +173,66 @@ static int mdns_do_tiebreak(DnsResourceKey *key, DnsAnswer *answer, DnsPacket *p
         return 0;
 }
 
+static bool mdns_should_reply_using_unicast(DnsPacket *p) {
+        /* Work out if we should respond using multicast or unicast. */
+
+        /* The query had a QU bit set, RFC 6762, section 5.4 */
+        if (p->wants_unicast_reply)
+                return true;
+
+        /* The query was a legacy "one-shot mDNS query", RFC 6762, sections 5.1 and 6.7 */
+        if (p->sender_port != MDNS_PORT)
+                return true;
+
+        /* The query was a "direct unicast query", RFC 6762, section 5.5 */
+        switch (p->family) {
+        case AF_INET:
+                if (!in_addr_equal(AF_INET, &p->destination, (union in_addr_union *) &MDNS_MULTICAST_IPV4_ADDRESS))
```

Comment:
`in4_addr_equal(&p->destination.in, &MDNS_MULTICAST_IPV4_ADDRESS)` is shorter and can do without cast